pacsman/dicom_interface.py

```diff
@@ -132,3 +142,30 @@ def fetch_thumbnail(self, series_id):
         :return: A path to a PNG file on success, None if not found
         """
         raise NotImplementedError
+
+    @staticmethod
+    def build_patient_result(result, ds, additional_tags=None):
+        patient_id = getattr_required(ds, 'PatientID')
+        study_instance_uid = getattr_required(ds, 'StudyInstanceUID')
+
+        if len(result) == 0:
```

Comment:
So if there was no pre-existing dataset for the patient, `result` would be a plain `Dataset` with no data from the defaultdict? And that has a "length" of 0? Clarifying the type of `result` would help here.

Reply:
Exactly.  This is the ugliest part of how `build_patient_result` works.  It is ugly, because there is this implicit assumption that the first time the function is called the dataset will be empty.  I don't think this is obvious to the caller.  I considered a couple other approaches:

1. Pass the result in as an optional keyword argument; its absence would indicate that this is the first time everything is called
2. Allow the `result` to be `None`, instead of an empty `Dataset`
3. Be clever, and grab each attribute from the `ds` if its "StudyDate" is newer or if the attribute is empty.  Indirectly this would

Perhaps it would be worth changing to 1 or 2?  3 seems more complicated, and potentially a performance issue.  BUT 3 has the advantage that it would more appropriately handle name changes or situations where some `additional_tags` are present in some DICOM dataset but not others.  I.e., the "result" would be built up from the intersection of the existing data instead of just from the first dataset.

Reply:
How about this:
Rename the method to `update_patient_result`.
Have it return `result`
Have `None` be the initial value.
Then this code both works and does what it looks like it does:
```
patient_result = None
for dataset in datasets:
    patient_result = update_patient_result(patient_result, dataset)
```
The `update` part of the name is a hint that `patient_result` is mutated.

Reply:
Agree, confusing to both mutate and return a value. If we drop that aspect, then starting with an empty result seems best.